### DISCUSSION TITLE
Anonymous and local interface/abstract types implementations

### DISCUSSION THREAD

**Discussion by Opiumtm** [dotnet/csharplang#738] [General]
It would be good to support true local interface (or abstract types) implementations, as it's supported in Java.

```C#
public interface ISomething
{
    void Do(int arg);
    void Do2(int arg1, int arg2);
}

void SomeMethod()
{
    ISomething a;
    Func<int, ISomething> factory;

    int result = 0;

    // "Anonymous" interface implementation
    // Same syntax as for anonymous types, but as it's assigned to interface variable,
    // compiler "know" what methods anonymous type should implement
    a = new {
        void Do(int arg) {
            // you shouldn't declare method as public
            result = arg; // result variable is captured as closure!
        }
        void Do2(int arg1, int arg2) {    
            result = arg1 + arg2; // result variable is captured as closure!
        }
    };    

    // Factory function. Anonymous interface implementation can be declared as lambda-function
    factory = (x) => new {
        void Do(int arg) {
            result = arg + x; // x lambda argument is presented as hidden field in anonymous class
        }
        void Do2(int arg1, int arg2) {    
            result = arg1 + arg2 + x; // x lambda argument is presented as hidden field in anonymous class
        }
    };

    a = x(5);

    // "Local" interface implementation inside method code
    class LocalImplementation : ISomething
    {
        void Do(int arg) {    
            result = arg; // result variable is captured as closure!
        }
        void Do2(int arg1, int arg2) {    
            result = arg1 + arg2; // result variable is captured as closure!
        }
    }

    a = new LocalImplementation();
}
```

**Comment by CyrusNajmabadi** [comment]
This looks like a dupe of https://github.com/dotnet/roslyn/issues/13  Is there a difference here between this proposal and that one?

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Worth tracking here anyway since this repo now relates to language changes?

I do think that anonymous and/or local implementations would be a nice feature and it's something that I use frequently enough in Java.  Delegates/lambdas are great but can only represent a single action.  If I'm building out the implementation of a fluent API it just adds to the boilerplate to then have to define a nested type and manually capture the appropriate data.

For the anonymous implementation I'd much prefer the Java-style syntax where the type being instantiated is the interface name:

```cs
var a = new ISomething {
    void Do(int arg) { ... }
    void Do2(int arg1, int arg2) { ... }
};
```

Of course going down this path opens numerous cans of worms, such as solving scope resolution (`ParentClass.this`), other nested members, initializers, constructors, whether it can be used to anonymously derive from any arbitrary type, etc.

**Comment by jnm2** [comment]
@CyrusNajmabadi It's such a juicy and immediately useful feature that I still can't believe you aren't the first to champion it. So weird.

**Comment by CyrusNajmabadi** [comment]
:)    It's very uninteresting to me (and i came from a Java background).

**Comment by Opiumtm** [comment]
@HaloFour I agree that java-like syntax is better. And, of course, it's make things more clear (you see what interface was implemented). But I think, it should be optional. Explicit interfere declaration can be omitted if left part of assignment isn't `var`, but a well-known type.

>solving scope resolution (ParentClass.this)

`this` should always reference to anonymous/local implementation (i.e. to the class where method is declared). Reason: it doesn't break existing code in case of copy/paste or refactorings. `this` of outer class could be captured as closure if you write `var parentObj = this;` at the outer method scope. It's very easy to capture outer class instance that way.

>other nested members

Declared anonymous/local class should always "win" in case of any ambiguity. Reason same as above - to minimize compiler errors after you do copy/paste. And you would do copy/paste when you would convert internal or private nested class to local/anonymous.

>initializers

Should be allowed as in regular classes - i.e. you can declare default property value as you do in regular class. Developer also can "initialize" property values with values from captured closures.

```C#
var v = "test string";

IInterface a = new {
    string StringProperty { get; set;} = v; // we set up default value from captured closure
}
```

>constructors

Anonymous class should have only parameterless constructor. Constructors with parameters shouldn't be allowed. Compiler would emit constructor with parameters behind the scene (to pass closure object and lambda parameters from factory function), but it's implementation details that developer should not know. Developer should be allowed to create only parameterless constructor. If base class (in the case of anonymous implementation of abstract class) haven't parameterless constructor, developer could pass captured closure values or lambda parameters to base class constructor. Compiler would care of correct values passing.

```C#
var v = "value";

AbstractParent a = new {
    this() // use "this" keyword for constructor name
      :base(v) // pass captured value to base class constructor
    { 
    }
};
```

>whether it can be used to anonymously derive from any arbitrary type

Definitely yes! Anonymous implementation is the equivalent of private nested class. Just with captured closures and mangled compiler-generated type name.

**Comment by DavidArno** [comment]
I can see two use-cases for this:

1. It replaces the need to declare a private type in one part of a file and reference it in another. Effectively, it's that type declaration and usage all rolled into one. 
2. It solves that age-old problem of creating an anonymous type in eg a linq expression, and then using that type elsewhere in the code. The following:
```cs
var result = from thing in things
             where thing.Foo == 1
             select new { thing.Bar, thing.Baz };
```
Becomes:
```cs
var result = from thing in things
             where thing.Foo == 1
             select new IThing { thing.Bar, thing.Baz };
```
and now `result` is an `IThing` and can easily be passed around one's code.

**Comment by Opiumtm** [comment]
@DavidArno one major benefit of local/anonymous classes is closure support. It can save **lots** of lines of code. Built-in ability to get and update local variables values without even single additional line of code. Closures is major reason why lambda functions and local functions are so powerful. It's natural to reuse same power with anonymous/local classes.

**Comment by alrz** [comment]
I think local type declarations are more appropriate here,
```cs
IFoo M() {
  class Local : IFoo { .. }
  return new Local();
}
```
pros: needs a lot less syntax changes, supports closures, attributes, etc. more readable overall.
cons: not anonymous.


PS: This `new IThing { thing.Bar, thing.Baz };` is entirely something else though. it should have been:
```cs
new IThing {
  public Bar Bar => thing.Bar;
  public Baz Baz => thing.Baz;
};
```

**Comment by DavidArno** [comment]
> PS: This new IThing { thing.Bar, thing.Baz }; is entirely something else though. it should have been:

I'd see it as a mashup of existing anonymous type features and what's proposed in the OP. 

For the line:
```cs
var x = new { thing.Bar, thing.Baz };
```
The resultant anonymous type has `Bar` and `Baz` properties.

So for the line:
```cs
var x = new IThing { thing.Bar, thing.Baz };
```
Then, as long as the signature of `IThing` matches, eg:
```
public interface IThing
{
    Bar Bar { get; }
    Baz Baz { get; }
}
```
then the anonymous type can implement that interface. This could then be extended by the ideas in this proposal, eg if `IThing` were:
```cs
public interface IThing
{
    Bar Bar { get; }
    Baz Baz { get; }
    int Do(int index);
}
```
Then the type declaration could be:
```cs
var x = new IThing 
        { 
            thing.Bar, 
            thing.Baz;
            int Do(int index) => Bar[index];
        };
```
And if the properties don't match, then the usual anonymous type property matching can be used:
```cs
var x = new IThing 
        { 
            Bar = someOtherThing.FooBar, 
            Baz = someOtherThing.FooBaz;
            int Do(int index) => Bar[index];
        };
```

**Comment by jnm2** [comment]
Local types would mollify me.

**Comment by tuespetre** [comment]
I have long wanted something like this; I agree with @HaloFour and @DavidArno's sketches that the instantiation should include the interface type's identifier, e.g. `var foo = new IFoo { ... }` rather than `IFoo foo = new { ... }`

**Comment by Opiumtm** [comment]
@tuespetre 
I propose local types for that.

```C#
    // "Local" interface implementation inside method code
    class LocalImplementation : ISomething
    {
        void Do(int arg) {    
            result = arg; // result variable is captured as closure!
        }
        void Do2(int arg1, int arg2) {    
            result = arg1 + arg2; // result variable is captured as closure!
        }
    }
```

Sometimes it's excessive to declare interface explicitly when left side of assignment is well-known.

```C#
ISomeInterface a = new {
  // interface members
}
```
-- for example

Of course, you can write `new Func<int, int>(x => x*x)` instead of just `x => x*x` when function type on left side is well-known and there is no ambiguity. But it's pointless in most cases.

Let's give developer freedom to choose when explicitly declare interface/base class type of anonymous implementation and when omit it.

**Comment by DavidArno** [comment]
@tuespetre,

At the risk of appearing contrary, I should point out that I'd be happy with either syntax (ideally both):
```cs
var foo = new IFoo { ... }
IFoo foo = new { ... }
```
I just tend to use `var ...` so would naturally gravitate to the first version, but I can see benefits to the latter syntax too.

**Comment by tuespetre** [comment]
Good points.

Maybe there is some interplay with the shapes proposal to be considered here too. :thinking:

**Comment by Joe4evr** [comment]
I could swear I commented about this, but maybe I just forgot to click the button.

Anyway, `IFoo foo = new { ... }` as a syntax option would come naturally if/when #100 gets implemented. So by elimination, `var foo = new IFoo { ... }` should be the one to get the primary focus for this feature.

**Comment by tuespetre** [comment]
@Joe4evr oh, nice! I hadn't looked at that one yet.

**Comment by Thaina** [comment]
I support this but wonder if it need to belong in this repo or roslyn?

**Comment by FANMixco** [comment]
I hope this will be implemented soon. This is highly important for proper Android support.

**Comment by jnm2** [comment]
A slightly different proposal by @tmat is at https://github.com/dotnet/csharplang/discussions/6049.
